<a href="https://colab.research.google.com/github/SamyySwift/Hamoye_Internship/blob/master/Quiz_on_energy_consumption.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Problem Definition: 
Given the dataset, Appliances Energy Consumption,  we are to develop a multiple linear regression model that can predict new energy consumption of an appliance when given new test data. To answer some questions, we will need to normalize the dataset using the MinMaxScaler after removing the following columns: [“date”, “lights”]. The target variable is “Appliances”. Use a 70-30 train-test set split with a  random state of 42 (for reproducibility).

## Importing Libraries

In [150]:
# importing required libraries for data manipulation and data wrangling

import pandas as pd
import matplotlib.pyplot  as plt
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler


# Importing required libraries for modelling and model evaluation

from sklearn.linear_model import Ridge, Lasso, LinearRegression
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

## Loading the Dataset

In [73]:
# loading the dataset
df = pd.read_csv('energydata_complete.csv')
df.head()

date  Appliances  lights  ...  Tdewpoint        rv1        rv2
0  2016-01-11 17:00:00          60      30  ...        5.3  13.275433  13.275433
1  2016-01-11 17:10:00          60      30  ...        5.2  18.606195  18.606195
2  2016-01-11 17:20:00          50      30  ...        5.1  28.642668  28.642668
3  2016-01-11 17:30:00          50      40  ...        5.0  45.410389  45.410389
4  2016-01-11 17:40:00          60      40  ...        4.9  10.084097  10.084097

[5 rows x 29 columns]

## Descriptive summary of the dataset

In [74]:
# descriptive summary of the data

df.describe()

Appliances        lights  ...           rv1           rv2
count  19735.000000  19735.000000  ...  19735.000000  19735.000000
mean      97.694958      3.801875  ...     24.988033     24.988033
std      102.524891      7.935988  ...     14.496634     14.496634
min       10.000000      0.000000  ...      0.005322      0.005322
25%       50.000000      0.000000  ...     12.497889     12.497889
50%       60.000000      0.000000  ...     24.897653     24.897653
75%      100.000000      0.000000  ...     37.583769     37.583769
max     1080.000000     70.000000  ...     49.996530     49.996530

[8 rows x 28 columns]

## Data Wrangling

In [75]:
# dropping the lights column since there are more zero entries as seen from the descriptive statistics  above
df.drop(['lights'], axis=1,inplace=True)

# dropping the date column since its not a time series problem
df.drop(['date'], axis=1, inplace=True)

#Data Preprocessing

In [76]:
# Applying MinMaxScaler to scale the features column 

# instantiate the scaler
scaler = MinMaxScaler()

# normalizing the features
normalized_features = pd.DataFrame(scaler.fit_transform(df), columns=df.columns)

In [77]:
# splitting data into features(predictors) and target(response) variables

# predictors
features = normalized_features.drop(['Appliances'], axis=1)

# target variable
target = normalized_features['Appliances']

In [78]:
# splitting data into training and testing set

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size = 0.3, random_state = 42)

In [79]:
# create a dictionary of different algorithms 
models = {'Ridge': Ridge(),
        'Lasso': Lasso(),
        'LinearRegression':LinearRegression()
        }


In [80]:
# helper funtion to compute the r2_score, RSS, RMSE, MAE, MSE on the testing set

def compute_score(models, X_train, X_test, y_train, y_test):
    """
    This function iteratively goes through all the models defined in the dictionary and 
    computes the r2_score, MSE, MAE and RMSE.
    
    Parameters: model, training set(X_train), test_set(X_test), train_labels(y_train), and test_labels(y_test).
    
    Returns: This funtion returns a dataFrame containing calculations of each models.
    """
    
    # store properties of each model
    model_properties = []
    
    # loop through the dictionary of models
    for reg_name, regressor in models.items():
        # empty dict for storing properties of each regression models
        reg_dict = {}
        # store the name of each model
        reg_dict['Name'] = reg_name
        # fit the regressor model
        regressor.fit(X_train, y_train)
        # compute the RSS
        reg_dict['RSS'] = round(np.sum(np.square(y_test - regressor.predict(X_test))), 2)
        # compute the r2_score
        reg_dict['r2_score'] = round(r2_score(y_test, regressor.predict(X_test)), 2)
        # compute the MAE
        reg_dict['MAE'] = round(mean_absolute_error(y_test, regressor.predict(X_test)), 2)
        #compute the mean_squared_error
        reg_dict['MSE'] = round(mean_squared_error(y_test, regressor.predict(X_test)), 3)
        # compute the RMSE
        reg_dict['RMSE'] = round(np.sqrt(mean_squared_error(y_test, regressor.predict(X_test))), 3)
        # append the properties of a each regressor to the model_properties list after every iteration
        model_properties.append(reg_dict)
     
    # create a dataframe with a list of all the model properties
    summary_df = pd.DataFrame(model_properties)
    
        
    return summary_df
        
        

In [81]:
# execute the function
compute_score(models, X_train, X_test, y_train, y_test)

Name    RSS  r2_score   MAE    MSE   RMSE
0             Ridge  45.42      0.15  0.05  0.008  0.088
1             Lasso  53.28     -0.00  0.06  0.009  0.095
2  LinearRegression  45.35      0.15  0.05  0.008  0.088

# Quiz Questions

###Question 1 : In the Linear regression, L2 regularization is equivalent to imposing a:

###Answer: Gaussian prior.

###Question 2: Cross validation:

###Answer: Is quaranteed to prevent overfitting.

###Question 3 : Ridge Regression:

###Answer: Reduces variance at the expence of higher bias.

###Question 4: In the different terms of the bias-tradeoff, which of the following is substantially more harmful to the test error than the training error?

###Answer: Variance.

###Question 5: What can you use to find the best fit line for Linear Regression?

###Answer: Least Square Error.(The linear Regression uses the least square methode as its cost funtion adn aims to reduce this cost funtion i.e reducing the distance between the actual point and the line of best fit.)

###Question 6: Which of the following is true about outliers in linear regression?

###Answer: Linear regression is sensitive to outliers. (The slope or gradient of the regression line will change due to outliers in most of the cases hence Linear Regression is definately sensitive to outliers.)

###Question 7: How many coefficients do you need to estimate a simple linear regression model(One, independent variable)?

###Answer: 2. (y = mx + c) where m and c are the coefficients of regression.

###Question 8: Adding more bias functions in a linear model:

###Answer: Decreases model bias.

###Question 9: A best fit line relating X and Y has a R-squared value of 0.75. How do I interpret this information?

###Answer: 75% of the variance in Y is explained by X.

###Question 10: The Lasso can be interpreted as least-squares regression where:

###Answer: Weights are regularized with the L1 norm.

###Question 11: Which of these assumptions of Linear Regression?

###Answer: Multivariate normality

###Question 12: From the dataset, fit a linear model on the relationship between the temperature in the living room in Celsius (x = T2) and the temperature outside the building (y = T6). What is the R^2 value in two D.P?

###Answer: 0.64


In [94]:
# Obtaining the Training set of the T2 and T6 columns from the normalised data
train_x = X_train[['T2']]
train_y = X_train[['T6']]

# Obtaining the Testing set of the T2 and T6 column from the normalised data  
test_x = X_test[['T2']]
test_y = X_test[['T6']]

# Instantiate the LinearRegression model
model = LinearRegression()

# Fit the model with training set
model.fit(train_x, train_y)

# Make predictions on test set
prediction = model.predict(test_x)


#Compute the r2_score on test set
R2_score = round(r2_score(test_y, prediction), 2)
R2_score

0.64

###Question 13: What is the mean absolute error

###Answer: 0.05


In [96]:
# Lets instantiate the general LinearRegression Model to answer questions 13 - 17

# Instantiate the LinearRegression Model
lin_reg = LinearRegression()

# fit the model with training set
lin_reg.fit(X_train, y_train)

# Make predictions on testing set
y_pred = lin_reg.predict(X_test)

In [97]:
# Computing the Mean_absolute_error
MAE = round(mean_absolute_error(y_test, y_pred), 2)

MAE

0.05

###Question 14: What is the Residual Sum of Squares(in two decimal places)

###Answer: 45.35


In [100]:
# Computing the Residual Sum of Squares in 2D.p
RSS = round(np.sum(np.square(y_test - y_pred)), 2)

RSS

45.35

###Question 15: What is the Root Mean Suared Error(in three decimal places)

###Answer: 0.088

In [103]:
RMSE = round(np.sqrt(mean_squared_error(y_test, y_pred)), 3)

RMSE

0.088

###Question 16: What is the Coefficient of Determination(in two decimal places)

###Answer: 0.15

In [104]:
# Computing the r2_score
R2_score = r2_score(y_test, y_pred)
round(R2_score, 2)

0.15

###Question 17: Obtain the feature weights from your linear model above. Which features have the lowest and highest weights respectively?

###Answer: We can see from the Table below that the features with the lowest and highest weights are RH_2, RH_1 respectively.

In [148]:
def get_weights_df(model, feat, col_name):
  #this function returns the weight of every feature
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

# Execute the get_weight function and store in dataframe
linear_weights_df = get_weights_df(lin_reg, X_train, 'Linear_weight')

linear_weights_df

Features  Linear_weight
0          RH_2      -0.456698
1         T_out      -0.321860
2            T2      -0.236178
3            T9      -0.189941
4          RH_8      -0.157595
5        RH_out      -0.077671
6          RH_7      -0.044614
7          RH_9      -0.039800
8            T5      -0.015657
9            T1      -0.003281
10          rv1       0.000770
11          rv2       0.000770
12  Press_mm_hg       0.006839
13           T7       0.010319
14   Visibility       0.012307
15         RH_5       0.016006
16         RH_4       0.026386
17           T4       0.028981
18    Windspeed       0.029183
19         RH_6       0.038049
20         RH_3       0.096048
21           T8       0.101995
22    Tdewpoint       0.117758
23           T6       0.236425
24           T3       0.290627
25         RH_1       0.553547

###Question 18: Train a Ridge Regression model with an alpha value of 0.4. Is there any change to the root mean squared error(RMSE) when evaluated on the test set?

###Answer: From the calculations below, there are no differences, hence the answer is No.

In [109]:
# Instantuate ridge model with default aplha value
ridge = Ridge()

# Fitting the ridge model with training set
ridge.fit(X_train, y_train)

# making predictions on test set
pred = ridge.predict(X_test)

ridge_rmse = round(np.sqrt(mean_squared_error(y_test, pred)), 3)

print(f'The RMSE score for the rigde model with default aplha value when evaluated on the test set is: {ridge_rmse}')


# Instantiate another ridge model with aplha value set to 0.4
ridge_reg = Ridge(alpha=0.4)

# fit on traing set
ridge_reg.fit(X_train, y_train)

# Make new predictions on test set
new_pred = ridge_reg.predict(X_test)

ridge_rmse_new = round(np.sqrt(mean_squared_error(y_test, new_pred)), 3)

print(f'The RMSE score for the rigde model with default aplha value when evaluated on the test set is: {ridge_rmse_new}')

print('There is no change in the RMSE scores')

The RMSE score for the rigde model with default aplha value when evaluated on the test set is: 0.088
The RMSE score for the rigde model with default aplha value when evaluated on the test set is: 0.088
There is no change in the RMSE scores


###Question 19: Train a Lasso regression model with an aplha value of 0.001 and obtain the new feature weights with it. How many of the features have non-zero feature weights?

###Answer: 4



In [127]:
# Instatiate a lasso regressor with alpha value of 0.001
lasso_reg = Lasso(alpha=0.001)

# Fit the lasso model with training set
lasso_reg.fit(X_train, y_train)

# Make predictions on test set
lasso_preds = lasso_reg.predict(X_test)

# defining a function to get weights
def get_weights_df(model, feat, col_name):
  #this function returns the weight of every feature
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

lasso_weights_df = get_weights_df(lasso_reg, X_train, 'Lasso_weight')

# Lets obtain the dataFrame with non-zero feature weights
non_zero_weights = lasso_weights_df[lasso_weights_df['Lasso_weight'] != 0]

# view non_zero dataframe
print(non_zero_weights)

print('')

# print the total number of non_zero weights
print(f'The total features with non-zero feature weights are equal to: {len(non_zero_weights)}')

     Features  Lasso_weight
0      RH_out     -0.049557
1        RH_8     -0.000110
24  Windspeed      0.002912
25       RH_1      0.017880

The total features with non-zero feature weights are equal to: 4


###Question 20: What is the new RMSE with Lasso Regression (in 3 decimal places)

###Answer: 0.094

In [149]:
# Compute the new RMSE with lasso regression
RMSE_new = round(np.sqrt(mean_squared_error(y_test, lasso_preds)), 3)

RMSE_new

0.094